### PR TITLE
Add elasticsearch reindex events

### PR DESCRIPTION
**Pull request by nlegoff**
First work on trigger elasticsearch indexation events
- [x] suppression d'un record
- [x] ajout d'un record
- [x] edition d'un record (caption, status)
- [x] changement de collection d'un record
- [x] substitution d'un record
- [x] creation d'une subdef
- [x] le champ documentaire est désigné comme titre
- [x] re-nommage d'une collection
- [x] suppression d'un champs
- [x] re-nommage de champs
- [x] fusion de champs
- [x] suppression d'un status
- [x] re-nommage d'un status

Missing thesaurus indexation events when user operation are done on thesaurus

@mdarse any thought ont this ?

### CODE REVIEW

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        $databox = $event->getDatabox();
+
+        if ($this->isIndexationOrdered($event, $databox)) {
+            return;
+        }
+
+        $bulk = $this->getBulk();
+
+        $this->recordIndexer->populateIndexOnDatabox($bulk, $databox);
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: This should be done asynchronously

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        $collection = $event->getCollection();
+        $databox = $collection->get_databox();
+
+        if ($this->isIndexationOrdered($event, $databox, $collection)) {
+            return;
+        }
+
+        $bulk = $this->getBulk();
+
+        $this->recordIndexer->populateIndexOnCollection($bulk, $collection);
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: This should be done asynchronously

**Comment by mdarse** [comment]
I don't have enough time fully look at all of this, but there are many events types, are all of them really needed?

Also, do we already have some kind of messaging service already set up?

File: lib/Alchemy/Phrasea/SearchEngine/Elastic/Indexer/RecordIndexer.php [hunk 3]
```diff
@@ -81,18 +112,26 @@ public function indexSingleRecord(\record_adapter $record_adapter, $indexName)
...
+            'type'  => self::TYPE_RECORD,
+            'id'    => $record->get_serialize_key()
+        ]);
+    }
+
     private function findLinkedConcepts($structure, array $record)
     {
         $client = $this->elasticSearchEngine->getClient();
         $searchParams['index'] = $this->elasticSearchEngine->getIndexName();
-        $searchParams['type']  = TermIndexer::TYPE_NAME;
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/SearchEngine/Elastic/Indexer/RecordIndexer.php]: revert

File: lib/Alchemy/Phrasea/SearchEngine/Elastic/Search/Query.php [hunk 4]
```diff
@@ -30,6 +30,7 @@ public function isFullTextOnly()
 
     public function getElasticsearchQuery($fields = array())
     {
+        $fields = array('_all');
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/SearchEngine/Elastic/Search/Query.php]: revert

File: lib/classes/databox/field.php [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<?php
+	<?php
```
Comment (nlegoff) [inline lib/classes/databox/field.php]: revert

**Comment by nlegoff** [comment]
@mdarse All event types are needed because they involve a reindexing process. I listed them in PR description.

For asynchronous tasks, we have our task manager for the moment. it's a "kind of messaging service".

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            PhraseaEvents::INDEX_COLLECTION => ['onIndexCollection', -24],
+            PhraseaEvents::INDEX_RECORD => ['onIndexRecord', -48],
+        ];
+    }
+
+    public function onIndexDatabox(DataboxIndexEvent $event)
+    {
+        $databox = $event->getDatabox();
+
+        if ($this->isIndexationOrdered($event, $databox)) {
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: This is kind of broken because only the first indexation process will be taken into account. For example if we have this use case
```php
$record->setStatus($status);
 // triggers indexations
$app['dispatcher']->dispatch(PhraseaEvents::RECORD_CHANGE_STATUS, new ChangeStatusEvent($record));

$record->setMetadata($metadata);
// triggers indexation
$app['dispatcher']->dispatch(PhraseaEvents::RECORD_CHANGE_METADATA, new ChangeMetadataEvent($record));
```

Only the last indexation order process must be fired.

**Comment by mdarse** [comment]
What I mean is, do we need such granularity in events? We may fire more generic events in multiple places. For example by replacing record events like `ChangeCollectionEvent`, `ChangeMetadataEvent`, `ChangeOriginalNameEvent` and `ChangeStatusEvent` by a simpler `RecordChangeEvent`.

Eventually, `RecordChangeEvent` can contain information about what property exactly changed, if we want to do more optimization in the future on the index side.

**Comment by nlegoff** [comment]
I prefer to keep a high granularity to keep full control of what happened. This will be useful later on to clean the record_adapter class (cache invalidation for example) and also to have precise extension points for Phraseanet plugins.

We have a lot this kind of code where in the same controller we call
```php
$record->setMetdata($metadata);
$record->setStatus($status)
...
// and so on
```
It will be hard in those cases to keep track of what changed if we have only one ```RecordChangeEvent```.

After speaking with @jygaulier about the messaging system to do the indexation asynchronously, we came to a point where we do not want to introduce a new way to do it like ```rabbitmq``` or something else for now. 

We prefer to keep the old system, which present many avantages (concurrency (multi-editing, events that trigger a data box indexation vs event that trigger a single record indexation), knowledge from the tech team) where we just raise a status bit if the record need to be reindexed (the first four status are reserved so we can use them), then a task grab those records periodically and (bulk) update them in ES. 

We will use two of these status one to tell that record need to be reindexed and an other one which tell us that the current record is being reindexed so a user can safely edit a record while it is being reindexed.

**Comment by mdarse** [comment]
Keeping the old system of marking records dirty with status bits seems a good idea to me. Introducing a messaging system can be done later if really needed.
The only issue I find here is with the polling the indexing worker *needs* to do, which is hackish and inefficient.
Anyway, we can find mean to notify the worker before triggering an heavy query on the DB. Maybe by polling the filesystem for a small marker file, it will be more efficient. But is there not a task manager right here that can be used for that purpose?

**Comment by nlegoff** [comment]
Yes we can use the task manager for that purpose.

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        $connection = $collection->get_connection();
+
+        // set 'to_index' fag where 'to_index' flag is not already set
+        // for given collection
+        // @todo sql query this should not be done right here
+        $sql = <<<SQL
+        UPDATE record
+        SET jeton = (jeton | :token)
+        WHERE coll_id = :coll_id
+        AND (jeton & (1 << (log2(:token))) = 0)
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: (jeton & :token) do the same

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        $databox = $event->getDatabox();
+        $connection = $databox->get_connection();
+
+        // set 'to_index' fag where 'to_index' flag is not already set
+        // for given databox
+        // @todo sql query this should not be done right here
+        $sql = <<<SQL
+        UPDATE record
+        SET jeton = (jeton | :token)
+        WHERE (jeton & (1 << (log2(:token))) = 0)
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: (jeton & :token) do the same

File: lib/Alchemy/Phrasea/Media/SubdefGenerator.php [hunk 9]
```diff
@@ -39,44 +40,42 @@ public function __construct(Application $app, Alchemyst $alchemyst, Filesystem $
 
     public function generateSubdefs(\record_adapter $record, array $wanted_subdefs = null)
     {
-        if (null === $subdefs = $record->get_databox()->get_subdef_structure()->getSubdefGroup($record->get_type())) {
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Media/SubdefGenerator.php]: revert this changes it is unnecessary for now

File: lib/Alchemy/Phrasea/Media/SubdefSubstituer.php [hunk 10]
```diff
@@ -24,12 +26,13 @@ class SubdefSubstituer
     private $fs;
     private $mediavorus;
 
-    public function __construct(Application $app, Filesystem $fs, Alchemyst $alchemyst, MediaVorus $mediavorus)
+    public function __construct(Application $app,Filesystem $fs, Alchemyst $alchemyst, MediaVorus $mediavorus, EventDispatcherInterface $dispatcher)
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Media/SubdefSubstituer.php]: revert space

File: lib/classes/databox/field.php [hunk 11]
```diff
@@ -422,7 +422,7 @@ public function save()
         $stmt->closeCursor();
 
         if ($this->renamed) {
-            caption_field::rename_all_metadatas($this);
+            caption_field::rename_all_metadatas($this->app['dispatcher'], $this);
```
Comment (nlegoff) [inline lib/classes/databox/field.php]: rename_all_metadata

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        $record = $event->getRecord();
+        if (self::$recordsToRemove->contains($record)) {
+            return;
+        }
+        self::$recordsToRemove->attach($record);
+    }
+
+    public function doFlushIndex()
+    {
+        if (self::$recordsToAdd->count() === 0 && self::$recordsToRemove->count() === 0 && self::$recordsToUpdate->count() === 0) {
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: unecessary

File: lib/Alchemy/Phrasea/Metadata/PhraseanetMetadataSetter.php [hunk 13]
```diff
@@ -93,6 +101,8 @@ public function replaceMetadata($metadataCollection, \record_adapter $record)
 
         if (count($metas) > 0) {
             $record->set_metadatas($metas, true);
+
+            $app['dispatcher']->dispatch(PhraseaEvents::RECORD_CHANGE_METADATA, new ChangeMetadataEvent($record));
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Metadata/PhraseanetMetadataSetter.php]: $this->dispatcher

File: lib/Alchemy/Phrasea/SearchEngine/Elastic/Fetcher/AbstractRecordFetcher.php [hunk 14]
```diff
@@ -9,79 +9,75 @@
...
         $this->addMetadataToRecords($records);
-        $this->addSubdefsToRecord($records);
+        $this->addSubDefinitionsToRecord($records);
 
         // Hydrate records
         foreach ($records as $key => $record) {
             $records[$key] = $this->hydrate($record);
         }
 
+        if (is_callable($this->postFetch)) {
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/SearchEngine/Elastic/Fetcher/AbstractRecordFetcher.php]: maybe use event dispatcher here instead of this tricks

File: lib/Alchemy/Phrasea/Core/PhraseaEvents.php [hunk 15]
```diff
@@ -42,4 +41,27 @@
     const BRIDGE_UPLOAD_FAILURE = 'bridge.upload-failure';
     const EXPORT_MAIL_FAILURE = 'export.mail-failure';
     const EXPORT_CREATE = 'export.create';
+
```
Comment (nmaillat) [inline lib/Alchemy/Phrasea/Core/PhraseaEvents.php]: add 
status_change_label 
event

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            PhraseaEvents::INDEX_DATABOX => ['onIndexDatabox', -12],
+            PhraseaEvents::INDEX_COLLECTION => ['onIndexCollection', -24],
+            PhraseaEvents::INDEX_NEW_RECORD => ['onIndexNewRecord', -48],
+            PhraseaEvents::INDEX_UPDATE_RECORD => ['onIndexUpdateRecord', -48],
+            PhraseaEvents::INDEX_REMOVE_RECORD => ['onIndexRemoveRecord', -48],
+            KernelEvents::FINISH_REQUEST => 'doFlushIndex',
```
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: KERNEL_TERMINATE no ?

File: lib/Alchemy/Phrasea/Core/PhraseaEvents.php (hunk 15 above)
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/PhraseaEvents.php]: done

**Comment by nlegoff** [comment]
This PR looks quit good now, ready for any review!

File: lib/Alchemy/Phrasea/Core/Event/CollectionEvent/CollectionEvent.php [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\CollectionEvent;
+
+use Symfony\Component\EventDispatcher\Event as SfEvent;
+
+class CollectionEvent extends SfEvent
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/CollectionEvent/CollectionEvent.php]: No need for `as SfEvent`, just extend `Event` no?
This should also be an abstract class.

File: lib/Alchemy/Phrasea/Core/Event/DataboxEvent/DataboxEvent.php [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\DataboxEvent;
+
+use Symfony\Component\EventDispatcher\Event as SfEvent;
+
+class DataboxEvent extends SfEvent
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/DataboxEvent/DataboxEvent.php]: No need for `as SfEvent`, just extend `Event` no?
This should also be an abstract class.

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/BuildSubdefEvent.php [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/BuildSubdefEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeOriginalNameEvent.php [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeOriginalNameEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeMetadataEvent.php [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeMetadataEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeCollectionEvent.php [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeCollectionEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeStatusEvent.php [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/ChangeStatusEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/CreateRecordEvent.php [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/CreateRecordEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/CreateStoryEvent.php [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/CreateStoryEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/DeleteRecordEvent.php [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/DeleteRecordEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/DeleteStoryEvent.php [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/DeleteStoryEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/RecordEvent.php [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Symfony\Component\EventDispatcher\Event as SfEvent;
+
+class RecordEvent extends SfEvent
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/RecordEvent.php]: No need for `as SfEvent`, just extend `Event` no?
This should also be an abstract class.

File: lib/Alchemy/Phrasea/Core/Event/RecordEvent/SubstituteRecordEvent.php [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) 2005-2014 Alchemy
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Alchemy\Phrasea\Core\Event\RecordEvent;
+
+use Alchemy\Phrasea\Model\Entities\Basket;
+use Symfony\Component\EventDispatcher\Event as SfEvent;
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/RecordEvent/SubstituteRecordEvent.php]: Not used

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/CollectionSubscriber.php [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+            PhraseaEvents::COLLECTION_CHANGE_NAME => 'onChangeName',
+        ];
+    }
+
+    public function onChangeName(ChangeNameEvent $event)
+    {
+        $collection = $event->getCollection();
+
+        $dispatcher = $event->getDispatcher();
+        $dispatcher->dispatch(PhraseaEvents::INDEX_COLLECTION, new CollectionIndexEvent($collection));
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/CollectionSubscriber.php]: I don't like much this part, events should not be used like command IMO. An event is about something that occurred in the past.
The code executing the index process is the best place to know on which events it have to react. Proxying `onChangeName` to `indexCollection` doesn't add value. In fact, we are losing the information of what happened.
This is the indexer which should eventually dispatch an `indexCollection` event when it is doing that. Else why not just have a method call?

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+        ];
+    }
+
+    public function onUpdateField(UpdateStructureFieldEvent $event)
+    {
+        $databox = $event->getDatabox();
+        $field = $event->getField();
+        $data = $event->getData();
+
+        if ((bool) $field->get_thumbtitle() !== (bool) $data['thumbtitle']) {
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: What does this do? It is not a string?

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+            $dispatcher = $event->getDispatcher();
+            $dispatcher->dispatch(PhraseaEvents::INDEX_DATABOX, new DataboxIndexEvent($databox));
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: Same remarks than in on `CollectionSubscriber.php`

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    }
+
+    public function onDeleteField(UpdateStructureFieldEvent $event)
+    {
+        $databox = $event->getDatabox();
+
+        $dispatcher = $event->getDispatcher();
+        $dispatcher->dispatch(PhraseaEvents::INDEX_DATABOX, new DataboxIndexEvent($databox));
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: Same remarks than in on `CollectionSubscriber.php`

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public function onUpdateStatus(UpdateStatusEvent $event)
+    {
+        $databox = $event->getDatabox();
+
+        $dispatcher = $event->getDispatcher();
+        $dispatcher->dispatch(PhraseaEvents::INDEX_DATABOX, new DataboxIndexEvent($databox));
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: Same remarks than in on `CollectionSubscriber.php`

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public function onDeleteStatus(DeleteStatusEvent $event)
+    {
+        $databox = $event->getDatabox();
+
+        $dispatcher = $event->getDispatcher();
+        $dispatcher->dispatch(PhraseaEvents::INDEX_DATABOX, new DataboxIndexEvent($databox));
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: Same remarks than on `CollectionSubscriber.php`

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php [hunk 36]
```diff
@@ -0,0 +1,194 @@
...
+use Alchemy\Phrasea\SearchEngine\Elastic\Fetcher\RecordFetcher;
+use Alchemy\Phrasea\SearchEngine\Elastic\Fetcher\RecordPoolFetcher;
+use Alchemy\Phrasea\SearchEngine\Elastic\Indexer\RecordIndexer;
+use Alchemy\Phrasea\SearchEngine\Elastic\Indexer\TermIndexer;
+use Alchemy\Phrasea\SearchEngine\Elastic\RecordHelper;
+use Elasticsearch\Client;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+class ElasticSearchSubscriber implements EventSubscriberInterface
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/ElasticsearchSubscriber.php]: It's the responsibility of the indexer component to do many of the work done in this class. A subscriber should be just about doing some plumbing work.
As a proof of matter, this class is not easily testable. It's not possible to check the business logic without magic-dancing with the event dispatcher. It's a little code smell sign IMO.

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/RecordSubscriber.php [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+
+    public function onChangeStatus(ChangeStatusEvent $event)
+    {
+        $dispatcher = $event->getDispatcher();
+        $dispatcher->dispatch(PhraseaEvents::INDEX_UPDATE_RECORD, new RecordIndexEvent($event->getRecord()));
+    }
+
+    public function onBuildSubDef(BuildSubDefEvent $event)
+    {
+        if (in_array($event->getSubDefName(), ['thumbnail', 'thumbnailgif', 'preview'])) {
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/RecordSubscriber.php]: This is business logic, it shouldn't be there.

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/StorySubscriber.php [hunk 38]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Alchemy\Phrasea\Core\Event\Subscriber;
+
+use Alchemy\Phrasea\Core\Event\RecordEvent\CreateStoryEvent;
+use Alchemy\Phrasea\Core\Event\RecordEvent\DeleteStoryEvent;
+use Alchemy\Phrasea\Core\PhraseaEvents;
+use Alchemy\Phrasea\SearchEngine\Elastic\Indexer\RecordIndexer;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class StorySubscriber implements EventSubscriberInterface
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/StorySubscriber.php]: I like this one! :)

File: lib/Alchemy/Phrasea/Core/Provider/PhraseaEventServiceProvider.php [hunk 39]
```diff
@@ -42,6 +47,27 @@ public function register(Application $app)
         $app['phraseanet.content-negotiation-subscriber'] = $app->share(function (Application $app) {
             return new ContentNegotiationSubscriber($app);
         });
+        $app['phraseanet.record-subscriber'] = $app->share(function (Application $app) {
```
Comment (mdarse) [inline lib/Alchemy/Phrasea/Core/Provider/PhraseaEventServiceProvider.php]: I don't think it's a good idea mix and match subscribers of various type together : we already have a service provider for the search engine, since the engine depends on the firing of these event there is little reasons to put subscribers here.

File: lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php (hunk 31 above)
Comment (nlegoff) [inline lib/Alchemy/Phrasea/Core/Event/Subscriber/DataboxSubscriber.php]: It just checks if user has change the property 'title' of a field, if yes we must reindex the data box because it changes the title of all records

**Comment by mdarse** [comment]
Follow up in #28.
